src/camera.rs

```diff
@@ -18,23 +18,59 @@ fn spawn(mut commands: Commands) {
 
 pub fn update(
     keys: Res<ButtonInput<KeyCode>>,
-    mut query: Query<(Ref<OrthographicProjection>, &mut Transform), With<Camera2d>>,
+    mut query: Query<(&OrthographicProjection, &mut Transform), With<Camera2d>>,
     time: Res<Time>,
     tilemaps: Res<Assets<Map>>,
     tilemap_query: Query<&Handle<Map>>,
+    window_query: Query<&Window>,
 ) {
+    let Ok((projection, mut transform)) = query.get_single_mut() else {
+        return;
+    };
+
+    let min_speed = 250.;
+    let max_speed = 500.;
+
     let x = keys.any_pressed([KeyCode::ArrowRight, KeyCode::KeyD]) as i8
         - keys.any_pressed([KeyCode::ArrowLeft, KeyCode::KeyA]) as i8;
     let y = keys.any_pressed([KeyCode::ArrowUp, KeyCode::KeyW]) as i8
         - keys.any_pressed([KeyCode::ArrowDown, KeyCode::KeyS]) as i8;
     let dir = Vec2::new(x as f32, y as f32).normalize_or_zero();
 
-    let Ok((projection, mut transform)) = query.get_single_mut() else {
+    if dir != Vec2::ZERO {
+        let speed = if keys.any_pressed([KeyCode::ShiftLeft, KeyCode::ShiftRight]) {
+            max_speed
+        } else {
+            min_speed
+        };
+
+        transform.translation += dir.extend(0.) * time.delta_seconds() * speed;
+    }
+
+    let Ok(window) = window_query.get_single() else {
         return;
     };
 
-    if dir == Vec2::ZERO && !projection.is_changed() {
-        return;
+    if let Some(cursor_position) = window.cursor_position() {
+        fn map_range(from: (f32, f32), to: (f32, f32), value: f32) -> f32 {
```

Comment:
I think this is provided by glam: <https://docs.rs/glam/0.27.0/glam/trait.FloatExt.html#tymethod.remap>

Reply:
Thanks, I didn't find this function